producer/writer/consumer_writer.go

```diff
@@ -0,0 +1,148 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/generated/proto/msgpb"
+	"github.com/m3db/m3msg/protocol/proto"
+	"github.com/m3db/m3x/log"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+type consumerWriter interface {
+	// Write writes the marshaler, it should be thread safe.
+	Write(m proto.Marshaler) error
+
+	// Init initializes the consumer writer.
+	Init()
+
+	// Close closes the consumer writer.
+	Close()
+}
+
+type consumerWriterMetrics struct {
+	ackError    tally.Counter
+	decodeError tally.Counter
+	encodeError tally.Counter
+}
+
+func newConsumerWriterMetrics(scope tally.Scope) consumerWriterMetrics {
+	return consumerWriterMetrics{
+		ackError:    scope.Counter("ack-error"),
+		decodeError: scope.Counter("decode-error"),
+		encodeError: scope.Counter("encode-error"),
+	}
+}
+
+type consumerWriterImpl struct {
+	// encodeLock controls the access to the encode function.
+	encodeLock sync.Mutex
+	encdec     proto.EncodeDecoder
+	router     ackRouter
+	opts       Options
+	logger     log.Logger
+
+	c      *retryableConnection
+	closed *atomic.Bool
+	wg     sync.WaitGroup
+	m      consumerWriterMetrics
+}
+
+// TODO: Remove the nolint comment after adding usage of this function.
+// nolint: deadcode
+func newConsumerWriter(
+	addr string,
+	router ackRouter,
+	opts Options,
+) consumerWriter {
+	if opts == nil {
+		opts = NewOptions()
+	}
+	conn := newRetryableConnection(addr, opts)
+	return &consumerWriterImpl{
+		encdec: proto.NewEncodeDecoder(conn, opts.EncodeDecoderOptions()),
+		router: router,
+		opts:   opts,
+		logger: opts.InstrumentOptions().Logger(),
+		c:      conn,
+		closed: atomic.NewBool(false),
+		m:      newConsumerWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+}
+
+// Write should fail fast so that the write could be tried on other
+// consumer writers that are sharing the data.
+func (w *consumerWriterImpl) Write(m proto.Marshaler) error {
+	w.encodeLock.Lock()
+	err := w.encdec.Encode(m)
+	w.encodeLock.Unlock()
+	if err != nil {
+		w.c.SignalReset()
+		w.m.encodeError.Inc(1)
+	}
+	return err
+}
+
+func (w *consumerWriterImpl) Init() {
+	w.c.Init()
+
+	w.wg.Add(1)
+	go func() {
```

Comment:
nit: move `w.wg.Done` to `readAcksForever` and simply call `go w.readAcksForever()` here.

Reply:
Sure it does come at the cost of an extra lambda allocation but that's fine.